hadoop-hdfs-project/hadoop-hdfs-rbf/src/main/java/org/apache/hadoop/hdfs/server/federation/router/RouterClientProtocol.java

```diff
@@ -194,6 +202,18 @@ public class RouterClientProtocol implements ClientProtocol {
     this.routerCacheAdmin = new RouterCacheAdmin(rpcServer);
     this.securityManager = rpcServer.getRouterSecurityManager();
     this.rbfRename = new RouterFederationRename(rpcServer, conf);
+
+    this.crsNameservicesCache = CacheBuilder.newBuilder()
```

Comment:
why do we decide to introduce a cache here? isNameSapceObserverREadEligible() does not even need to contact NN, right? It just checks for local configs.

Reply:
To avoid computing the set of eligibleNameservices again and again. It may be a premature optimization though.
I'm also not sure how the equality check in the cache compares to the previous code, performance wise.
@mkuchenbecker thoughts?

Reply:
I misread the implementation: I thought the key was a nameservice and then the value is whether that nameservice has an observer node or not. 

There are some inconsistency in what we are trying to achieve here. In one hand, we are leveraging static config in determining whether a nameservice has an observerNode or not. On the other hand, we are leveraging a dynamic cache with a size of 1 here, assuming the set of namespaces would actually be changed. It is better to be consistent here, either based on fixed static info or dynamic. 

If we are leveraging static config in determining observerNode, we can assume we are using a static fixed set of nameservices as well. We can just do the check once and then assume the nameservices won't change. Or is this proposal too naive? 😂

Reply:
The set off all active nameservices changes as well. For example all the namenodes in ns0 can be decommissioned while the routers are still up.

Reply:
I do think the code without the cache is much easier to understand and the performance difference is negligible. I'll remove the caching code and we can add it back in as a separate commit if necessary.